README.md
# symmetric-cipher.js
This is a lightweight library meant for symmetric encryption and decryption of strings. It is written for express-js to add support for encrypted cookies, but will also work with Node.js.

This library currently only allows users to use 'AES-256-CBC' symmetric algorithm. On encryption, a random IV is computed and prepended to the ciphertext with `:` as the separator. This library uses the in-built Node.js `crypto` library under the hood.

256-bit (32-byte) keys are required for encryption or decryption. To avoid imposing strict requirements, this library automatically generates a sha256 hash of the key if it is not the required length. However, it is not recommended to use short keys like 'super secret key', though they will still work.

See the [Security section](#security) for tips on generating keys.

## Installation
NPM
```bash
npm install symmetric-cipher.js
```
Yarn
```bash
yarn add symmetric-cipher.js
```

## Compatibility
This library is a commonJS library compatible with Node >= 0.10


## Usage
```javascript
var cipher = require('symmetric-cipher.js');

// Encrypt
cipher.encrypt('my value', 'super secret key');
// rSgTZLgXrFlpAOQkUuTgmQ==:wb/6vFECOGRCSA3DSVLpjw==

// Decrypt
cipher.decrypt('rSgTZLgXrFlpAOQkUuTgmQ==:wb/6vFECOGRCSA3DSVLpjw==', 'super secret key');
// my value
```

ES Modules
```javascript
import cipher from 'symmetric-cipher.js';

// Encrypt
cipher.encrypt('my value', 'super secret key');
// rSgTZLgXrFlpAOQkUuTgmQ==:wb/6vFECOGRCSA3DSVLpjw==

// Decrypt
cipher.decrypt('rSgTZLgXrFlpAOQkUuTgmQ==:wb/6vFECOGRCSA3DSVLpjw==', 'super secret key');
// my value
```
Named imports *may* also work for Node >= 14.13.0.

## Security
The key used for encryption and decryption should be a long, random string. The longer the key, the more secure the encryption. A 32-byte (256-bit) key in this case is sufficient. The key should be stored securely.

### Generating a base64 key (Recommended)
```bash
node
crypto.randomBytes(32).toString('base64')
```
Example: `ktFZQO5JDd4wjwPB2U6bawZKUTGbgOfw6up0wFtjsVk=`

### Generating a hex key
```bash
node
crypto.randomBytes(32).toString('hex')
```
Example: `37f3ce113b34e789ae76ee238c8aa349a6b163e81c283144bffb9c17328360d1`


**NOTE:** For keys generated by the above commands, don't forget to specify the encoding. See the [Examples section](#examples) below.

Store the key securely. For Node.js, you can store it in a `.env` file and use [dotenv](https://github.com/motdotla/dotenv) library to load it into your application. Don't store your key in version control. Add `.env` to `.gitignore`.

## API
### **decodeKey(key, keyEncoding)**
This is a helper function that converts a key to a Buffer object (binary). It leaves Buffer objects intact. It is used internally by the encrypt and decrypt functions.
- `key`: (string | Buffer) The secret used for encryption or decryption. If it is a string, it is converted from the specified encoding to binary.
- `keyEncoding` (string) The encoding of the key. Default is `utf8`. Others include `base64`, `hex`, `ascii`. For full details see [crypto's BufferEncoding](https://nodejs.org/api/buffer.html#buffers-and-character-encodings).
### **encrypt(plainText, key, keyEncoding)**

- `plaintext` (string) The string to encrypt
- `key`: (string | Buffer) The secret used for encryption or decryption. If it is a string, it is converted from the specified encoding to binary.
- `keyEncoding` (string) The encoding of the key. Default is `utf8`. Others include `base64`, `hex`, `ascii`. For full details see [crypto's BufferEncoding](https://nodejs.org/api/buffer.html#buffers-and-character-encodings).

Output is in the format {24-character-iv}:{ciphertext} in base64 encoding.

### **decrypt(cipherText, key, keyEncoding)**
Expected input is in the format {24-character-iv}:{ciphertext} in base64 encoding.

- `cipherText` (string) The cipher to decrypt
- `key`: (string | Buffer) The secret used for encryption or decryption. If it is a string, it is converted from the specified encoding to binary.
- `keyEncoding` (string) The encoding of the key. Default is 'utf8'. Others include 'base64', 'hex', 'ascii'. For full details see [crypto's BufferEncoding](https://nodejs.org/api/buffer.html#buffers-and-character-encodings).

Output is the original plaintext.

## Examples
Base64 key (Recommended)
```javascript
var cipher = require('symmetric-cipher.js');

cipher.encrypt('my value', 'P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')

cipher.decrypt('y3axcNW9eA6ggJNYWotsUQ==:k+YSdJ1MBZct/0bOtQ/zvA==', 'P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')
```

Key as Buffer object (binary)
```javascript
var cipher = require('symmetric-cipher.js');

var key = cipher.decodeKey('P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')

cipher.encrypt('my value', key)

cipher.decrypt('y3axcNW9eA6ggJNYWotsUQ==:k+YSdJ1MBZct/0bOtQ/zvA==', key)
```

Hex key
```javascript
var cipher = require('symmetric-cipher.js');

cipher.encrypt('my value', '3fac1504251a027465981346fb5b0d57d398e4df4a03253a4c7d1926e40e9907', 'hex')

cipher.decrypt('y3axcNW9eA6ggJNYWotsUQ==:k+YSdJ1MBZct/0bOtQ/zvA==', '3fac1504251a027465981346fb5b0d57d398e4df4a03253a4c7d1926e40e9907', 'hex')
```

## License
[MIT](./LICENSE)

## Author
[Fadhili Njagi](https://github.com/FadhiliNjagi)


HISTORY.md
2.1.0-0
===
* Add support for keys with encoding prefix e.g. `base64:P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=`
* Remove buffer-from dependency

2.0.0
===
* Test integration with other libraries. So far so good.

2.0.0-dev
===
* Add `decodeKey(key, keyEncoding)`
* Change cipher encoding format from hex to base64, to create shorter strings
* deps: buffer-from@1.1.2
  - Add compatibility with older node versions (>= 0.10) for `Buffer.from`
* Remove esm distribution. Library is now purely commonJS (es5)
* More thorough tests

1.0.0
===
* Stable API
* Cookie encryption and decryption
* Security guidelines for encryption keys


0.1.0 (deleted)
===
* Unstable API
* README typos

test.ts
import assert from 'assert'
import { encrypt, decrypt, decodeKey } from './index'
// eslint-disable-next-line n/no-deprecated-api
const bufferFrom = (typeof Buffer.from === 'function') ? Buffer.from : function (str: string, enc: BufferEncoding) { return new Buffer(str, enc) }

describe('decodeKey = (key, keyEncoding)', function () {
  it('should convert a string key to a buffer', function () {
    const key = 'super secret key'
    assert.ok(decodeKey(key) instanceof Buffer)
    const base64key = 'P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc='
    assert.ok(decodeKey(base64key, 'base64') instanceof Buffer)
    assert.ok(decodeKey(`base64:${base64key}`) instanceof Buffer)
    const hexKey = '3fac1504251a027465981346fb5b0d57d398e4df4a03253a4c7d1926e40e9907'
    assert.ok(decodeKey(hexKey, 'hex') instanceof Buffer)
    assert.ok(decodeKey(`hex:${hexKey}`) instanceof Buffer)
  })

  it('should leave a buffer key intact', function () {
    const key = bufferFrom('super secret key', 'utf8')
    assert.strictEqual(decodeKey(key), key)
  })
})

describe('decrypt(cipherText, key, keyEncoding)', function () {
  describe('when cipher format is invalid', function () {
    it('should throw an invalid format error', function () {
      assert.throws(function () {
        decrypt('foobar', 'P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=')
      }, /Invalid cipher text format. Expected {24-character-iv}:{cipher-text}/)
    })
  })

  describe('when key is a string', function () {
    it('should decrypt cipher text', function () {
      const plainText = 'foobar'
      const cipherText = 'kI7KX7DpxGiQko4k2hPkaQ==:wN39aymkUhx8KVenmijULw=='
      const secret = 'super secret key'
      assert.strictEqual(decrypt(cipherText, secret), plainText)
    })
  })

  describe('when key is a buffer', function () {
    it('should decrypt cipher text', function () {
      const plainText = 'foobar'
      const cipher = 'kI7KX7DpxGiQko4k2hPkaQ==:wN39aymkUhx8KVenmijULw=='
      const secret = bufferFrom('P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')
      assert.strictEqual(decrypt(cipher, secret), plainText)
    })
  })

  describe('when key encoding is specified as prefix', function () {
    it('should decrypt cipher text', function () {
      const plainText = 'foobar'
      const cipher = 'kI7KX7DpxGiQko4k2hPkaQ==:wN39aymkUhx8KVenmijULw=='
      const secret = 'base64:P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc='
      assert.strictEqual(decrypt(cipher, secret), plainText)
    })
  })
})

describe('encrypt(plainText, key, keyEncoding)', function () {
  describe('when key is a string', function () {
    it('should encrypt plain text', function () {
      const plainText = 'foobar'
      const secret = 'super secret key'
      const encrypted = encrypt(plainText, secret)
      assert.notStrictEqual(encrypted, plainText)
      assert.ok(/^[A-Za-z0-9+/=]{24}:[A-Za-z0-9+/=]+$/.test(encrypted))
      assert.strictEqual(decrypt(encrypted, secret), plainText)
    })
  })

  describe('when key is in different encoding', function () {
    it('should encrypt with the specified key encoding', function (done) {
      const plainText = 'foobar'
      const base64Secret = 'P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc='
      const encryptedWithBase64 = encrypt(plainText, base64Secret, 'base64')
      const encryptedWithUtf8 = encrypt(plainText, base64Secret, 'utf8')
      assert.strictEqual(decrypt(encryptedWithBase64, base64Secret, 'base64'), plainText)
      assert.throws(function () {
        decrypt(encryptedWithUtf8, base64Secret, 'base64')
      }, /bad decrypt/)
      assert.throws(function () {
        decrypt(encryptedWithBase64, base64Secret, 'utf8')
      }, /bad decrypt/)

      done()
    })
  })

  describe('when key is a buffer', function () {
    it('should encrypt plain text', function () {
      const plainText = 'foobar'
      let secretBuffer = bufferFrom('3fac1504251a027465981346fb5b0d57d398e4df4a03253a4c7d1926e40e9907', 'hex')
      let encrypted = encrypt(plainText, secretBuffer)
      assert.notStrictEqual(encrypted, plainText)
      assert.strictEqual(decrypt(encrypted, secretBuffer), plainText)

      secretBuffer = bufferFrom('P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')
      encrypted = encrypt(plainText, secretBuffer)
      assert.notStrictEqual(encrypted, plainText)
      assert.strictEqual(decrypt(encrypted, secretBuffer), plainText)
    })
  })

  describe('when key encoding is specified as prefix', function () {
    it('should encrypt plain text', function () {
      const plainText = 'foobar'
      let secretBuffer = 'hex:3fac1504251a027465981346fb5b0d57d398e4df4a03253a4c7d1926e40e9907'
      let encrypted = encrypt(plainText, secretBuffer)
      assert.notStrictEqual(encrypted, plainText)
      assert.strictEqual(decrypt(encrypted, secretBuffer), plainText)

      secretBuffer = 'base64:P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc='
      encrypted = encrypt(plainText, secretBuffer)
      assert.notStrictEqual(encrypted, plainText)
      assert.strictEqual(decrypt(encrypted, secretBuffer), plainText)
    })
  })

  describe('when key is not exactly 256 bytes long', function () {
    it('should encrypt plain text with sha256 of key', function () {
      const plainText = 'foobar'
      const shortSecret = 'super secret key'
      const sha256Secret = bufferFrom('P6wVBCUaAnRlmBNG+1sNV9OY5N9KAyU6TH0ZJuQOmQc=', 'base64')
      const encryptedWithShortSecret = encrypt(plainText, shortSecret)
      const encryptedWithSha256Secret = encrypt(plainText, sha256Secret)
      // Assert that the keys are interchangeable
      assert.strictEqual(decrypt(encryptedWithShortSecret, sha256Secret), decrypt(encryptedWithSha256Secret, shortSecret))
    })
  })
})


index.ts
/**
 * cipher-js
 * Copyright (c) 2023 Fadhili Njagi
 * MIT licensed
*/
import { createHash, randomBytes, createCipheriv, createDecipheriv } from 'crypto'
// eslint-disable-next-line n/no-deprecated-api
const bufferFrom = (typeof Buffer.from === 'function') ? Buffer.from : function (str: string, enc: BufferEncoding) { return new Buffer(str, enc) }

/**
 * Decodes a string to a Buffer Object. Leaves Buffer objects intact.
 *
 * @param key - the encryption key
 * @param keyEncoding - the encryption key's encoding scheme (default: `utf8`)
 * @return key
 * @api public
 */
export const decodeKey = (key: string | Buffer, keyEncoding: BufferEncoding = 'utf8'): Buffer => {
  if (!(key instanceof Buffer)) {
    const keyParts = key.split(':')
    if (keyParts[0] !== undefined && keyParts[1] !== undefined) {
      if (Buffer.isEncoding(keyParts[0])) {
        keyEncoding = keyParts[0]
        key = keyParts[1]
      }
    }
    key = bufferFrom(key, keyEncoding)
  }
  return key
}

/**
 * Adjusts keylength. If key is not exactly 256 bytes long, returns sha256 of key
 *
 * @param key - the encryption key
 * @return 256-bit key
 * @private
 */
const adjustKeyLength = (key: Buffer): Buffer => {
  if (key.length !== 32) {
    key = createHash('sha256').update(key).digest()
  }
  return key
}

/**
 * Encrypts a string using AES-256-CBC algorithm
 *
 * @param plainText - string to encrypt
 * @param key - the encryption key
 * @param keyEncoding - the encryption key's encoding scheme (default: `utf8`)
 * @return the encrypted value
 * @api public
 */
export const encrypt = (plainText: string, key: string | Buffer, keyEncoding: BufferEncoding = 'utf8'): string => {
  key = decodeKey(key, keyEncoding)
  key = adjustKeyLength(key)
  const iv = randomBytes(16)
  const cipher = createCipheriv('aes-256-cbc', key, iv)
  const encrypted = [iv.toString('base64'), ':', cipher.update(plainText, 'utf8', 'base64'), cipher.final('base64')].join('')
  return encrypted
}

/**
 * Decrypts a string using AES-256-CBC algorithm
 *
 * @param cipherText - cipher to decrypt
 * @param key - the encryption key
 * @param keyEncoding - the encryption key's encoding scheme (default: `utf8`)
 * @return the decrypted value
 * @api public
 */
export const decrypt = (cipherText: string, key: string | Buffer, keyEncoding: BufferEncoding = 'utf8'): string => {
  const textParts = cipherText.split(':')
  if (textParts[0] === undefined || textParts[1] === undefined) {
    throw new Error('Invalid cipher text format. Expected {24-character-iv}:{cipher-text}')
  }
  key = decodeKey(key, keyEncoding)
  key = adjustKeyLength(key)
  const iv = bufferFrom(textParts[0], 'base64')
  const encryptedText = bufferFrom(textParts[1], 'base64')
  const decipher = createDecipheriv('aes-256-cbc', key, iv)
  const decrypted = Buffer.concat([decipher.update(encryptedText), decipher.final()])
  return decrypted.toString()
}
